### PR TITLE
backends: restore shlex quoting of MESONINTROSPECT

### LINKED ISSUE #12148: Arguments in MESONINTROSPECT are escaped incosistently
**Describe the bug**
The arguments in the `MESONINTROSPECT` environment variable are escaped with the `join_args` function in `mesonbuild/utils/universal.py` for postconf scripts while they are escaped with `shlex.quote()` in the other situations.

`join_args` and `shlex.quote()` have different behaviors on Windows. `join_args` returns a string that is compatible with Win32 `CreateProcess` function, but `shlex.quote()` returns a string that is compatible with Python `shlex.split()` function. However, neither is compatible with both functions (e.g., `shlex.quote()` is not compatible with Win32 `CreateProcess`.)

The incompatibility of `join_args` and `shlex.split()` broke `test cases/common/139 mesonintrospect from scripts` and the breakage was detected with #12115. 

Here are possible ways to fix the inconsistency:
1. Use `shlex.quote()` instead of `join_args` to escape the arguments in `MESONINTROSPECT` for postconf scripts. This breaks post scripts that pass `MESONINTROSPECT` to Windows APIs.
2. ~Always use `join_args` to escape the arguments in `MESONINTROSPECT` and make `join_args` emit a string compatible with `shlex.split()`. More concretely, make it replace `\` with `\\`. This adds extra verbosity to `MESONINTROSPECT`.~ No, Windows treats `\\` literally.
3. Always use `join_args` to escape the arguments in `MESONINTROSPECT` and make `check_introspection.py` pass it to cmd on Windows. This abandons compatibility with `shlex.split()` and makes it harder to use `MESONINTROSPECT` in Python scripts.

@xclaesse suggested always using `join_args` and checking other `shlex.quote()` usage. I found the other usage is fine:

- `mesonbuild/coredata.py` uses the function to get `meson.build_options()` but the documentation says not to parse it:
https://github.com/mesonbuild/meson/blob/bde690b06e930020a0ec6ccaea7a76babf77dff5/docs/yaml/builtins/meson.yaml#L174
- `mesonbuild/mtest.py` uses the function if it's *not* on Windows. It returns a string as is on Windows and has "TODO" comment. The output is used only for logging:
https://github.com/mesonbuild/meson/blob/543e9ca0cf0c00d752bd723ec403e91b839bf9b4/mesonbuild/mtest.py#L229
- `mesonbuild/backend/ninjabackend.py` uses the function for GCC response file. GCC response file has its own parser so the compatibility with Windows APIs does not matter.
- `mesonbuild/modules/external_project.py` uses the function for environment variables like `CFLAGS`. Such variables should always have POSIX shell syntax since they are used by configure scripts and Makefile that use POSIX shell.

@eli-schwartz on the other hand suggests always using `shlex.quote()` since the usage of `join_args` is a breaking change introduced with commit 522392e7553823e6b3ac38cadc4fbee72eae9540.

I'd like to point out the documentation also says `MESONINTROSPECT` is compatible with `shlex.split()`:
https://github.com/mesonbuild/meson/blob/bde690b06e930020a0ec6ccaea7a76babf77dff5/docs/yaml/builtins/meson.yaml#L89

@xclaesse @eli-schwartz Taking all of the above into account, what do you think we should use for `MESONINTROSPECT`, `shlex.split()` or `join_args`?

**To Reproduce**
Run tests with #12115.

**Expected behavior**
The tests should pass.

**system parameters**
* Is this a [cross build](https://mesonbuild.com/Cross-compilation.html) or just a plain native build (for the same computer)? Native.
* what operating system (e.g. MacOS Catalina, Windows 10, CentOS 8.0, Ubuntu 18.04, etc.) Arch Linux ARM
* what Python version are you using e.g. 3.8.0 3.11.3
* what `meson --version` 1.2.1
* what `ninja --version` if it's a Ninja build 1.11.1

### PR DESCRIPTION
**Pull request by joukewitteveen**
The type of quoting was changed in 522392e to one that is suitable for use with `cmd.exe` on Windows. However, the documentation states that the type of quoting in `MESONINTROSPECT` is compatible with `shlex.split()` and elsewhere in the code, the same variable is still quoted with `shlex.quote()`. As mostly identified in #12148, there are a few choices:
1. Use `shlex.quote()` consistently and support Python but not `cmd.exe`.
2. Use `join_args` and support `cmd.exe` but not Python.
3. Use `join_args` and support splitting through the `mesonbuild` Python library.

This commit implements the first option and reverts part of 522392e.

Regression testing is implemented in #12115.

Fixes #12148

### CODE REVIEW

**Comment by joukewitteveen** [comment]
The `' '.join(shlex.quote(x) for x in ...)` code occurs quite often. Maybe we should make a `mesonlib` function for it as well? It would be a sibling of `join_args` and maybe prevent people like me from being tempted to 'clean up' and just use `join_args` in the future.

**Comment by eli-schwartz** [comment]
It's really just a reimplementation of python 3.8's `shlex.join()`. I think a suitable internal name would be `mesonlib.join_unix_args`, which we might or might not want to differentiate from `join_native_args` (by renaming join_args). Depends on whether we figure that potential contributors will understand by default that `join_args` is native by default and `join_unix_args` is for when you specifically want unix style, which does seem like a reasonable assumption.

**Comment by joukewitteveen** [comment]
> `join_unix_args` is for when you specifically want unix style

We can maybe do a bit better here, because crucially this style will be easy to use from Python as well (not just from a Unix shell). That makes it sensible for use on Windows too (e.g. this whole `MESONINTROSPECT` business). At any rate, this can be done after merging the current pull request, which fixes a real bug.

File: mesonbuild/backend/backends.py [hunk 1]
```diff
@@ -1598,22 +1598,23 @@ def eval_custom_target_command(
         cmd = [i.replace('\\', '/') for i in cmd]
         return inputs, outputs, cmd
 
+    def get_introspect_command(self) -> str:
+        return ' '.join([shlex.quote(x) for x in self.environment.get_build_command() + ['introspect']])
```
Comment (eli-schwartz) [inline mesonbuild/backend/backends.py:1602]: join should accept a generator expression, no need for a list comprehension here.

**Review by eli-schwartz — CHANGES_REQUESTED** [review]

File: mesonbuild/backend/backends.py (hunk 1 above)
Comment (joukewitteveen) [reply mesonbuild/backend/backends.py:1602]: I agree and it is also how it is implemented in `shlex.join` (I thought that maybe there was some obscure size-hint allocation benefit from using a list). Yet, I chose to use the pattern that is used everywhere else in `meson` for consistency.

**Review by eli-schwartz — APPROVED** [review]

**Pull request merged**